packages/core/snapshot/snapshot.go

```diff
@@ -1,82 +1,150 @@
+		return nil, fmt.Errorf("fail to create snapshot file: %s", err)
 	}
 
-	if err = s.FromBytes(bytes); err != nil {
-		return errors.Errorf("failed to unmarshal Snapshot: %w", err)
+	outputWithMetadataProd := NewLedgerOutputWithMetadataProducer(t.Ledger)
+	committableEC, fullEpochIndex, err := t.Options.CommitmentFunc()
+	if err != nil {
+		return nil, err
 	}
 
-	return nil
-}
+	header := &ledger.SnapshotHeader{
+		FullEpochIndex: fullEpochIndex,
+		DiffEpochIndex: committableEC.EI(),
+		LatestECRecord: committableEC,
+	}
 
-func (s *Snapshot) FromBytes(bytes []byte) (err error) {
-	s.LedgerSnapshot = new(ledger.Snapshot)
-	_, err = serix.DefaultAPI.Decode(context.Background(), bytes, s.LedgerSnapshot)
+	err = StreamSnapshotDataTo(f, header, outputWithMetadataProd, nmgr.SnapshotEpochDiffs)
 	if err != nil {
-		return errors.Errorf("failed to read LedgerSnapshot: %w", err)
+		return nil, err
 	}
+	f.Close()
 
-	for _, output := range s.LedgerSnapshot.OutputsWithMetadata {
-		output.SetID(output.M.OutputID)
-		output.Output().SetID(output.M.OutputID)
-	}
+	return header, err
+}
 
-	for _, epochdiff := range s.LedgerSnapshot.EpochDiffs {
-		for _, spentOutput := range epochdiff.Spent() {
-			spentOutput.SetID(spentOutput.M.OutputID)
-			spentOutput.Output().SetID(spentOutput.M.OutputID)
-		}
-		for _, createdOutput := range epochdiff.Created() {
-			createdOutput.SetID(createdOutput.M.OutputID)
-			createdOutput.Output().SetID(createdOutput.M.OutputID)
-		}
+// LoadStreamableSnapshot loads a snapshot file from the given file path. Contents in a snapshot file
+// will not be written to a snapshot struct in case blowing up the memory, they should be proccessed in
+// consumer functions. To construct a snapshot struct from a file, use FromBytes([]byte).
+func LoadStreamableSnapshot(filePath string,
+	outputWithMetadataConsumer OutputWithMetadataConsumerFunc,
+	epochDiffsConsumer EpochDiffsConsumerFunc,
+	notarizationConsumer NotarizationConsumerFunc) (err error) {
+
+	f, err := os.Open(filePath)
+	defer f.Close()
+	if err != nil {
+		return fmt.Errorf("fail to open the snapshot file")
 	}
 
-	return nil
+	err = StreamSnapshotDataFrom(f, outputWithMetadataConsumer, epochDiffsConsumer, notarizationConsumer)
+
+	return
 }
 
+// Write a snapshot struct in to a file with the given file name.
 func (s *Snapshot) WriteFile(fileName string) (err error) {
-	if err = os.WriteFile(fileName, s.Bytes(), 0o644); err != nil {
+	data, err := s.Bytes()
+	if err != nil {
+		return err
+	}
+
+	if err = os.WriteFile(fileName, data, 0o644); err != nil {
 		return errors.Errorf("failed to write snapshot file %s: %w", fileName, err)
 	}
 
 	return nil
 }
 
 // Bytes returns a serialized version of the Snapshot.
-func (s *Snapshot) Bytes() (serialized []byte) {
-	return marshalutil.New().
-		WriteBytes(lo.PanicOnErr(serix.DefaultAPI.Encode(context.Background(), s.LedgerSnapshot))).
-		Bytes()
+func (s *Snapshot) Bytes() (serialized []byte, err error) {
```

Comment:
This method assumes a non-stremeable snapshot (as it loads the entirety of `s.LedgerSnapshot.OutputsWithMetadata` at once). Why do we need it?

Reply:
In tools/genesis-snapshot where we initialize a snapshot struct and serialize it directly to a file.

Reply:
Can we move it there then? It is better to remove all "non-streamable" snapshot reading / writing out of the main codebase.